apps/aecore/lib/aecore/chain/chainstate.ex

```diff
@@ -106,23 +106,51 @@ defmodule Aecore.Chain.Chainstate do
     }
   end
 
-  defp calculate_miner_reward_chain_state(txs, chainstate, block_height, miner) do
-    case miner do
-      @genesis_miner ->
-        chainstate
+  defp calculate_miner_reward_chain_state(chainstate, block_height) do
+    if block_height > GovernanceConstants.beneficiary_reward_lock_time() do
+      current_generation_beneficiary =
```

Comment:
How does this work? ~I don't think `Map.keys` guarantees returning in the order of adding.~ I think it would be better to to get the header of block at proper height.

Reply:
After further inspection - I see that this will be the height of beneficiary block not pubkey.
So:
1. This height could be calculated with some simple substraction
2. It's still better to get the header by height from Chain.Worker in case of forks.

Reply:
Ok, this would mean removing the cache in general, don't you think it would be quite expensive to get all of the generation micro blocks from the db and iterate over them to get the fees again?

Reply:
How would we know which fork we are getting blocks from if we are getting them by height? 

>  I think it would be better to to get the header of block at proper height.
What do you mean by this? Just to get the beneficiary key?

Reply:
@cytadela8 
https://github.com/aeternity/elixir-node/pull/798/files#diff-76c70afbd7cea50d60f10c30ae67b41cR555
We are querying the beneficiary pub key.

We are not checking whether we are applying the rewards for the correct fork (see comment above)